### PR TITLE
Changes for input xaml conversion

### PR DESCRIPTION
**Pull request by Sofiya-kumar**

### CODE REVIEW

File: DemoApp/DemoApp/Views/TMInputPage.xaml [hunk 1]
```diff
@@ -3,62 +3,73 @@
...
-                       TitleText="Success/Error Input"
-                       HelperText="Helper Text"
-                       Placeholder="Text Input"
-                       InputValidation="CustomValidateInput"
-                       Focused="modusInputFocused" />
+        <controls:CustomInput x:Name="rightIcon"
+                              TitleText="Right Icon"
+                              Placeholder="Text Input"
+                              RightIconSource="password_eye.png"
+                              RightIconCommand="{Binding ShowPasswordCommand}"
```
Comment (BAL4-R) [inline DemoApp/DemoApp/Views/TMInputPage.xaml:29]: The right icon command is not working now

File: Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,551 @@
...
+    /// <param name="bindable">Object</param>
+    /// <param name="oldValue">Old value</param>
+    /// <param name="newValue">New value</param>
+    private static void OnReadOnlyPropertyChanged(BindableObject bindable, object oldValue, object newValue)
+    {
+        if (bindable is CustomInput tmInput)
+        {
+            if ((bool)newValue)
+            {
+                tmInput.inputHelperLabel.Opacity = disabledOpacity;
```
Comment (BAL4-R) [inline Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs:423]: Opacity does not change for readonly

File: Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,551 @@
...
+        {
+            if ((bool)newValue)
+            {
+                SetBorderColor(customInput);
+            }
+            else
+            {
+                customInput.inputBorder.Stroke = ResourcesDictionary.ColorsDictionary(ColorsConstants.TrimbleGray);
+                customInput.inputBorder.StrokeThickness = 1;
+            }
```
Comment (BAL4-R) [inline Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs:545]: Need to set `disabledOpacity`

File: Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                                  Unfocused="InputBorderlessEntry_Unfocused"
+                                  Placeholder="{Binding Placeholder}"
+                                  Text="{Binding Text}"
+                                  IsPassword="{Binding IsPassword}"
+                                  IsReadOnly="{Binding IsReadOnly}"
+                                  TextColor="{Binding TextColor}"
+                                  IsEnabled="{Binding IsEnabled}"
+                                  FontAttributes="{Binding FontAttributes}"
+                                  PlaceholderColor="{Binding PlaceholderColor}"
+                                  CursorPosition="{Binding CursorPosition}"
```
Comment (BAL4-R) [inline Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml:57]: This is not needed, since we removed `CursorPosition` property

**Review by BAL4-R — CHANGES_REQUESTED** [review]

File: Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml [hunk 5]
```diff
@@ -0,0 +1,91 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<ContentView xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
```
Comment (rthanga1) [inline Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml:2]: Custom validation works in reverse. It shows `Invalid Text` for valid text and vice versa.
Screen Shot 2023-06-13 at 10 37 54 AM

File: Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml [hunk 6]
```diff
@@ -0,0 +1,91 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<ContentView xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
+             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
+             x:Class="Trimble.Modus.Components.Controls.CustomInput"
+             x:Name="tmInput"
+             xmlns:converters="clr-namespace:Trimble.Modus.Components.Converters"
+             xmlns:controls="clr-namespace:Trimble.Modus.Components">
+
+  <ContentView.Resources>
```
Comment (rthanga1) [inline Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml:9]: And the password eye icon also not working now. Previously it used to show/hide the text in the input field

File: Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs [hunk 7]
```diff
@@ -0,0 +1,551 @@
...
+
+public partial class CustomInput : ContentView
+{
+
+    #region Private Properties
+
+    private const double disabledOpacity = 0.4;
+
+    private ValidationResponse _validationResponse;
+    private CustomInput _customInput;
```
Comment (rthanga1) [inline Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs:16]: Can you explain the need of this reference? This just holds`this`.

File: Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs [hunk 8]
```diff
@@ -0,0 +1,551 @@
...
+    /// <summary>
+    /// Identifies the TextProperty bindable property. This property is used to display the property text
+    /// </summary>
+    public static readonly BindableProperty TextProperty =
+       BindableProperty.Create(nameof(Text), typeof(string), typeof(CustomInput), propertyChanged: OnTextChanged);
+
+    /// <summary>
+    /// Identifies the TextColor bindable property. This property is used to define the color of the entry text
+    /// </summary>
+    public static readonly BindableProperty TextColorProperty =
```
Comment (rthanga1) [inline Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs:31]: The user cannot change the TextColor of the Input. Do we have to expose this property now?

Comment (rthanga1) [reply Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs:31]: Same question for the Font and FontSize properties in this class.

File: Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs [hunk 9]
```diff
@@ -0,0 +1,551 @@
...
+    /// Delegate for the input validation function
+    /// </summary>
+    /// <param name="sender">Reference to the sender input control</param>
+    /// <returns>Tuple of bool and string, bool represent the result of the text validation and string represent the text to be displayed</returns>
+    public delegate Tuple<bool, string> InputValidationHandler(object sender);
+
+    ///// <summary>
+    ///// Public event handler to be invoked when text is changed
+    ///// </summary>
+    //public event EventHandler<TextChangedEventArgs> TextChanged;
```
Comment (rthanga1) [inline Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs:173]: Is this got removed now? We need a listener and a command for `TextChanged` event.

File: Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs [hunk 10]
```diff
@@ -0,0 +1,551 @@
...
+        {
+            SetBorderColor(customInput);
+        }
+    }
+
+    private void InputBorderlessEntry_Focused(object sender, FocusEventArgs e)
+    {
+        if (sender is BorderlessEntry)
+        {
+            SetBorderColor(_customInput);
```
Comment (rthanga1) [inline Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs:521]: We can just pass `this` right? Do we have really have to hold `this` in a variable?

**Review by rthanga1 — CHANGES_REQUESTED** [review]

File: Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs (hunk 8 above)
Comment (Sofiya-kumar) [reply Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs:31]: It was already in TMInput so I had added the same to this

File: Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs (hunk 10 above)
Comment (Sofiya-kumar) [reply Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs:522]: Replaced with this

File: Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml (hunk 5 above)
Comment (Sofiya-kumar) [reply Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml:2]: I had changed the validation as alphanumeric alone so its showing like that. Shall I change that validation regex?

File: Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml [hunk 11]
```diff
@@ -0,0 +1,91 @@
+<?xml version="1.0" encoding="utf-8" ?>
```
[the remark this answers is not in the document]
Comment (Sofiya-kumar) [reply Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml:1]: Sure Will do that

File: Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml (hunk 4 above)
Comment (Sofiya-kumar) [reply Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml]: Okay will remove it

File: Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml (hunk 6 above)
Comment (Sofiya-kumar) [reply Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml:9]: Changed and resolved the issue

File: DemoApp/DemoApp/Views/TMInputPage.xaml (hunk 1 above)
Comment (Sofiya-kumar) [reply DemoApp/DemoApp/Views/TMInputPage.xaml:29]: Resolved issue

File: Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs (hunk 3 above)
Comment (Sofiya-kumar) [reply Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs:546]: Added

File: Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs (hunk 7 above)
Comment (Sofiya-kumar) [reply Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs:16]: Replaced with this

File: Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs (hunk 9 above)
Comment (Sofiya-kumar) [reply Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs]: Added event changed

File: Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs (hunk 2 above)
Comment (Sofiya-kumar) [reply Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs]: Updated the opactiy

File: Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml (hunk 5 above)
Comment (rthanga1) [reply Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml:2]: Oh, I didn't know that. Can you update the helper text with what is a valid text? Like 'Enter a valid email' or in your case 'Enter only alphanumeric character'?

File: Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs (hunk 8 above)
Comment (Sofiya-kumar) [reply Trimble.Modus.Components/Controls/TMInput/CustomInput.xaml.cs]: Okay Removed those properties

File: DemoApp/DemoApp/Views/TMInputPage.xaml.cs [hunk 12]
```diff
@@ -1,6 +1,6 @@
 using DemoApp.ViewModels;
 using System.Text.RegularExpressions;
-using Trimble.Modus.Components;
+using Trimble.Modus.Components.Controls;
```
Comment (rthanga1) [inline DemoApp/DemoApp/Views/TMInputPage.xaml.cs:3]: I see you have moved the controls to the `Controls` folder, but please keep the namespace as it is. Users should use just `Trimble.Modus.Components`.

File: Trimble.Modus.Components/Controls/TMInput/TMInput.xaml.cs [hunk 13]
```diff
@@ -0,0 +1,484 @@
...
+            {
+                case ValidationResponse.Success:
+                    tmInput.inputBorder.Stroke = ResourcesDictionary.ColorsDictionary(ColorsConstants.Green);
+                    tmInput.inputHelperIcon.Source = ImageSource.FromFile(ImageConstants.TickImage);
+                    tmInput.inputBorder.StrokeThickness = 1;
+                    break;
+                case ValidationResponse.Error:
+                    tmInput.inputBorder.Stroke = ResourcesDictionary.ColorsDictionary(ColorsConstants.DangerRed);
+                    tmInput.inputHelperIcon.Source = ImageSource.FromFile(ImageConstants.ErrorImage);
+                    tmInput.inputBorder.StrokeThickness = 1;
```
Comment (BAL4-R) [inline Trimble.Modus.Components/Controls/TMInput/TMInput.xaml.cs:408]: Border thickness is 2 for valid and error state of the input

File: DemoApp/DemoApp/Views/TMModalPage.xaml [hunk 14]
```diff
@@ -3,17 +3,18 @@
...
              xmlns:Modus="clr-namespace:Trimble.Modus.Components;assembly=Trimble.Modus.Components"
+             xmlns:control="clr-namespace:Trimble.Modus.Components.Controls;assembly=Trimble.Modus.Components"
              Title="TMModalPage">
   <ScrollView>
     <VerticalStackLayout Spacing="5"
                          Padding="8">
       <Label Text="Modal"
              FontSize="Medium"
              FontAttributes="Bold" />
-      <Modus:TMInput x:Name="ModalTitle"
```
Comment (rthanga1) [inline DemoApp/DemoApp/Views/TMModalPage.xaml:13]: The name `Modus` was obvious. But `control` doesn't actually convey what it means.

File: Trimble.Modus.Components/Controls/BorderlessEntry/BorderlessEntry.cs [hunk 15]
```diff
@@ -2,7 +2,7 @@ namespace Trimble.Modus.Components;
 
 public class BorderlessEntry : Entry
 {
-    internal BorderlessEntry()
```
Comment (rthanga1) [inline Trimble.Modus.Components/Controls/BorderlessEntry/BorderlessEntry.cs:5]: Any reason for making this `public`? We don't want to expose this class to the outside world, right?

File: Trimble.Modus.Components/Controls/TMInput/TMInput.xaml [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+        </ImageButton>
+        <controls:BorderlessEntry x:Name="inputBorderlessEntry"
+                                  Grid.Column="1"
+                                  Focused="InputBorderlessEntry_Focused"
+                                  Unfocused="InputBorderlessEntry_Unfocused"
+                                  Placeholder="{Binding Placeholder}"
+                                  Text="{Binding Text}"
+                                  IsPassword="{Binding IsPassword}"
+                                  IsReadOnly="{Binding IsReadOnly}"
+                                  TextColor="{Binding TextColor}"
```
Comment (rthanga1) [inline Trimble.Modus.Components/Controls/TMInput/TMInput.xaml:69]: Can you remove the TextColor here? We don't have the `TextColor` in the .cs file and it throws a warning now

File: Trimble.Modus.Components/Controls/TMInput/TMInput.xaml.cs (hunk 13 above)
Comment (Sofiya-kumar) [reply Trimble.Modus.Components/Controls/TMInput/TMInput.xaml.cs]: Updated with 2

File: DemoApp/DemoApp/Views/TMModalPage.xaml (hunk 14 above)
Comment (Sofiya-kumar) [reply DemoApp/DemoApp/Views/TMModalPage.xaml:13]: Modified to modus

**Review by BAL4-R — APPROVED** [review]

File: Trimble.Modus.Components/Controls/TMInput/TMInput.xaml (hunk 16 above)
Comment (Sofiya-kumar) [reply Trimble.Modus.Components/Controls/TMInput/TMInput.xaml]: Removed textcolor

**Review by rthanga1 — APPROVED** [review]

**Pull request merged**